app/src/main/java/com/android/feedme/model/data/RecipeRepository.kt

```diff
@@ -107,32 +134,53 @@ class RecipeRepository(private val db: FirebaseFirestore) {
       onSuccess: (List<Recipe>, DocumentSnapshot?) -> Unit,
       onFailure: (Exception) -> Unit
   ) {
+    // Prefix search query
     var queryRef =
         db.collection(collectionPath)
-            .whereGreaterThanOrEqualTo("title", query)
-            .whereLessThan("title", query + "\uf8ff")
-            .limit(10) // Limit the number of documents fetched
+            .whereArrayContainsAny("searchItems", listOf(query.lowercase(Locale.ROOT)))
+            .limit(6) // Limit the number of documents fetched
 
     if (lastRecipe != null) {
       queryRef = queryRef.startAfter(lastRecipe)
     }
 
     queryRef
         .get()
-        .addOnSuccessListener {
-          it.documents.map { recipeMap ->
-            val data = recipeMap.data
-            if (data != null) {
-              val success = { recipe: Recipe? ->
-                onSuccess(listOfNotNull(recipe), it.documents.lastOrNull())
-              }
-              mapToRecipe(data, success, onFailure)
-            }
-          }
-        }
+        .addOnSuccessListener { addSuccessListener(it, onSuccess, onFailure) }
         .addOnFailureListener { onFailure(it) }
   }
 
+  /**
+   * A helper function that adds the recipes to the list of recipes
+   *
+   * @param snapshot: the snapshot of the query
+   * @param onSuccess: the callback function to be called on success
+   * @param onFailure: the callback function to be called on failure
+   */
```

Comment:
Nice way to avoid high code complexity ! Would love to have a bit more comments as the code is obscure, I have a hard time understanding what it does exactly

Reply:
Will add some comments 👍